AP_GPS: make receiver logging configurable through parameters

flyingthingsintothings opened the pull request:
Add parameters that allow users to configure logging from within their ground control station software.
The implementation is kept generic so all drivers that want this can implement it.
The following new parameters are added:

- `GPS_LOG_HZ`: Set frequency at which data is logged in the receiver
- `GPS_LOG_LEVEL`: Set level (Lite, Basic, Default or Full) of detail that should be logged
- `GPS_LOG_FORCE`: Set whether the driver should overwrite existing logging (Enabled) or add to it (Disabled)

flyingthingsintothings commented:
@SeptentrioGNSS

flyingthingsintothings commented on libraries/AP_GPS/AP_GPS.cpp:
```diff
@@ -427,6 +427,31 @@ const AP_Param::GroupInfo AP_GPS::var_info[] = {
+    // @DisplayName: Log frequency
+    // @Description: The frequency with which the receiver logs information.
+    // @Units: Hz
+    // @Values: 0:Off, 1:0.1Hz, 2:0.2Hz, 5:0.5Hz, 10:1Hz, 20:2Hz, 50:5Hz, 80:8Hz, 100:10Hz, 200:20Hz
```
The reasoning was to choose 1 for 0.1Hz as that is the lowest frequency a user would probably want. Choosing 1 for 1Hz would prevent adding a lower frequency in the future as there is no lower value than 1. I also tried to choose values that are available on common receivers.

Do you have a recommendation for how to change the parameter so it's more user friendly and generic? What should be the mapping between values (0.1Hz, 1Hz, 2Hz...) and underlying data (1, 10, 50...)? I would love any input on this approach.

amilcarlucas replied:
Just use floating point Hz.  And explain in the description that *the receiver will not guarantee that, it will only do a best effort to achieve it*

flyingthingsintothings replied:
Should the floating point values still be discrete (enum) or a continuous interval? For me it makes the most sense to use commonly available discrete values that are used on a best effort basis as most GNSS receivers use discrete frequency values for logging (100ms, 200ms, 1s...). A continuous interval could lead users to think that their receiver will use the exact value they choose, which could confuse them when they look at the actual value on their receiver.

Thank you for the help.

amilcarlucas replied:
Use a continuous floating point Hz. And explain in the description that the receiver will not guarantee that, it will only do a best effort to achieve it

flyingthingsintothings commented:
I updated the parameter to be a float instead. I also changed the logging command to set up incremental logging which is required to have correctly split logs across reboots. Tested the parameters in Mission Planner as well and they show up fine. Tested with and without overwrite, with frequency of 0Hz (off) and 1 and 2Hz and it worked fine. I still need to choose correct output blocks for each level so I will leave this as a draft for now.

SeptentrioGNSS commented:
could the setFileNaming command send not be named px4log but "ardup" or "log" as with incremental only the first 5 chars are used (right padded with _ if needed) followed by modulo-1000 counter

flyingthingsintothings commented:
@WickedShell Is this PR ready to merge or would it need further changes?

tridge commented:
@chiara-septentrio please test!
and if you can, do us a SITL simulator for SBF :-)

chiara-septentrio commented:
I tested the changes and made some of my own to make the flow of configuration work better. The program reacts correctly correctly to each parameters and I can see them in Mission Planner, but I can't see the description or set the parameters from Mission Planner (I changed the parameters in the code for testing). I would guess the parameters need to be accepted to the main code for it to be possible, am I correct @tridge ?

Georacer commented:
> I tested the changes and made some of my own to make the flow of configuration work better. The program reacts correctly correctly to each parameters and I can see them in Mission Planner, but I can't see the description or set the parameters from Mission Planner (I changed the parameters in the code for testing). I would guess the parameters need to be accepted to the main code for it to be possible, am I correct @tridge ?

It is correct that you will not be able to see the parameter description in Mission Planner. These need to be backed in MP compilation.
So you will not be able to have a drop-down menu and select it.
But you should be able to input the raw value directly.
![params](https://github.com/user-attachments/assets/95c29284-d117-458b-8f41-101b2cce903d)

chiara-septentrio commented:
I haven't made any headway on this. The parameters are displayed and when changed in mission planner, the change is kept in memory and displayed correctly. However, the gps is still configured as if the parameters were at default
